src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -96,6 +98,54 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument)
             codeDocument.SetDocumentIntermediateNode(document);
         }
 
+        private void InheritImportedDirectives(DocumentIntermediateNode document, string mainDocumentFilePath)
+        {
+            var visitor = new DirectiveVisitor();
+            visitor.VisitDocument(document);
+
+            var seenDirectives = new HashSet<DirectiveDescriptor>();
+            for (var i = visitor.Directives.Count - 1; i >= 0; i--)
+            {
+                var reference = visitor.Directives[i];
+                var directive = (DirectiveIntermediateNode)reference.Node;
+                var descriptor = directive.Descriptor;
+                var seenDirective = !seenDirectives.Add(descriptor);
+                var imported = directive.Source.HasValue && directive.Source.Value.FilePath != mainDocumentFilePath;
+
+                if (!imported)
+                {
+                    continue;
+                }
+
+                switch (descriptor.Kind)
+                {
+                    case DirectiveKind.SingleLine:
+                        if (seenDirective && descriptor.Usage == DirectiveUsage.FileScopedSinglyOccurring)
+                        {
+                            // This directive has been overridden, it should be removed from the document.
+
+                            break;
+                        }
+
+                        continue;
+                    case DirectiveKind.RazorBlock:
+                    case DirectiveKind.CodeBlock:
+                        if (descriptor.Usage == DirectiveUsage.FileScopedSinglyOccurring)
+                        {
+                            // A block directive cannot be imported.
+
+                            document.Diagnostics.Add(
+                                RazorDiagnosticFactory.CreateDirective_BlockDirectiveCannotBeImported(descriptor.Directive));
+                        }
+                        break;
```

Comment:
Make sure to have a default case here, even if it just throws and says unreachable. Then if we add more kinds then we know this code has to be touched